File: iosApp/Saat/Infrastructure/Persistence/Local/LocalDoaDataSource.swift
//
//  LocalDoaDataSource.swift
//  Saat
//
//  Created by Elmee on 25/06/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Foundation

internal final class LocalDoaDataSource: Sendable {
    private let dhikrSlugs: Set<String> = [
        "morning-dhikir",
        "evening-dhikir",
        "dhikir-after-salah",
        "sleep-dhikir"
    ]

    internal init() {}

    internal func getCatalog() async throws -> [DoaCatalogEntry] {
        let categories = loadCategories().data ?? []
        return categories.compactMap { cat in
            guard let slug = cat.slug, !slug.isEmpty else { return nil }
            let title = (cat.name != nil && !cat.name!.isEmpty) ? cat.name! : slug
            let kind: DoaCatalogKind = dhikrSlugs.contains(slug) ? .dhikr : .doa
            return DoaCatalogEntry(slug: slug, title: title, kind: kind)
        }
    }

    internal func getDailyDoas() async throws -> [DoaItem] {
        return loadDoaResponse(fileName: "daily.json").data ?? []
    }

    internal func getDoasBySlug(slug: String) async throws -> [DoaItem] {
        let file = "duas_\(slug).json"
        if assetExists(path: "Resources/doa/\(file)") || assetExists(path: file) {
            return loadDoaResponse(fileName: file).data ?? []
        } else if slug == "daily" {
            return try await getDailyDoas()
        } else {
            return []
        }
    }

    internal func getDhikrBySlug(slug: String) async throws -> [DhikrBundle] {
        let file = "dhikir_\(slug).json"
        guard let data = readAsset(path: file) else { return [] }
        do {
            let response = try JSONDecoder().decode(DhikrListResponse.self, from: data)
            return response.data ?? []
        } catch {
            print("Failed to decode Dhikr response for \(slug): \(error)")
            return []
        }
    }

    private func loadCategories() -> DoaCategoriesResponse {
        guard let data = readAsset(path: "categories.json") else {
            return DoaCategoriesResponse(success: false, message: nil, data: [])
        }
        do {
            return try JSONDecoder().decode(DoaCategoriesResponse.self, from: data)
        } catch {
            print("Failed to decode categories.json: \(error)")
            return DoaCategoriesResponse(success: false, message: nil, data: [])
        }
    }

    private func loadDoaResponse(fileName: String) -> DoaListResponse {
        guard let data = readAsset(path: fileName) else {
            return DoaListResponse(success: false, message: nil, data: [])
        }
        do {
            return try JSONDecoder().decode(DoaListResponse.self, from: data)
        } catch {
            print("Failed to decode Doa file \(fileName): \(error)")
            return DoaListResponse(success: false, message: nil, data: [])
        }
    }

    private func assetExists(path: String) -> Bool {
        let filename = (path as NSString).lastPathComponent
        let name = (filename as NSString).deletingPathExtension
        let ext = (filename as NSString).pathExtension
        return Bundle.main.url(forResource: name, withExtension: ext) != nil ||
            Bundle.main.url(forResource: name, withExtension: ext, subdirectory: "Resources/doa") != nil
    }

    private func readAsset(path: String) -> Data? {
        let filename = (path as NSString).lastPathComponent
        let name = (filename as NSString).deletingPathExtension
        let ext = (filename as NSString).pathExtension

        var url = Bundle.main.url(forResource: name, withExtension: ext)
        if url == nil {
            url = Bundle.main.url(forResource: name, withExtension: ext, subdirectory: "Resources/doa")
        }
        guard let fileUrl = url else { return nil }
        return try? Data(contentsOf: fileUrl)
    }
}

File: app/src/main/java/app/kamy/saatApp/infrastructure/local/LocalDoaDataSource.kt
package app.kamy.saatApp.infrastructure.local

import android.content.Context
import app.kamy.saatApp.domain.model.DhikrBundle
import app.kamy.saatApp.domain.model.DhikrListResponse
import app.kamy.saatApp.domain.model.DoaCatalogEntry
import app.kamy.saatApp.domain.model.DoaCatalogKind
import app.kamy.saatApp.domain.model.DoaCategoriesResponse
import app.kamy.saatApp.domain.model.DoaItem
import app.kamy.saatApp.domain.model.DoaListResponse
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

@Singleton
class LocalDoaDataSource @Inject constructor(
    @ApplicationContext private val context: Context,
    private val json: Json
) {
    private val dhikrSlugs = setOf(
        "morning-dhikir",
        "evening-dhikir",
        "dhikir-after-salah",
        "sleep-dhikir",
        "dzikiralmathuratkubro",
        "dzikiralmathuratsughro"
    )

    suspend fun getCatalog(): List<DoaCatalogEntry> = withContext(Dispatchers.IO) {
        val categories = loadCategories().data.orEmpty()
        categories.mapNotNull { cat ->
            val slug = cat.slug?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
            val title = cat.name?.takeIf { it.isNotBlank() } ?: slug
            val kind = if (slug in dhikrSlugs) DoaCatalogKind.DHIKR else DoaCatalogKind.DOA
            DoaCatalogEntry(slug = slug, title = title, kind = kind, nameData = cat.nameData)
        }
    }

    suspend fun getDailyDoas(): List<DoaItem> = withContext(Dispatchers.IO) {
        loadDoaResponse("daily.json").data.orEmpty()
    }

    suspend fun getDoasBySlug(slug: String): List<DoaItem> = withContext(Dispatchers.IO) {
        val file = "duas_$slug.json"
        if (assetExists("doa/$file")) {
            loadDoaResponse(file).data.orEmpty()
        } else if (slug == "daily") {
            getDailyDoas()
        } else {
            emptyList()
        }
    }

    suspend fun getDhikrBySlug(slug: String): List<DhikrBundle> = withContext(Dispatchers.IO) {
        val file = "dhikir_$slug.json"
        if (!assetExists("doa/$file")) return@withContext emptyList()
        readAsset("doa/$file").let { text ->
            json.decodeFromString(DhikrListResponse.serializer(), text).data.orEmpty()
        }
    }

    private fun loadCategories(): DoaCategoriesResponse =
        readAsset("doa/categories.json").let {
            json.decodeFromString(DoaCategoriesResponse.serializer(), it)
        }

    private fun loadDoaResponse(fileName: String): DoaListResponse =
        readAsset("doa/$fileName").let {
            json.decodeFromString(DoaListResponse.serializer(), it)
        }

    private fun assetExists(path: String): Boolean =
        runCatching { context.assets.open(path).close(); true }.getOrDefault(false)

    private fun readAsset(path: String): String =
        context.assets.open(path).bufferedReader().use { it.readText() }
}
